### PR TITLE
added stl read/write functions

### PR DESCRIPTION
**Pull request by nicola-giuliani**
I've added functions to read and write stl files using opencascades functions. Four tests check stl write to file, refinement using stl and NormalToMeshManifold and stl creation starting from iges files.

### CODE REVIEW

File: include/deal.II/opencascade/utilities.h [hunk 1]
```diff
@@ -128,6 +128,35 @@ namespace OpenCASCADE
...
+  /**
+   * Write the given topological shape into an STL file. In order to do so the
+   * shape must contain a mesh structure, the function checks if all the faces
+   * of the shape have an attached mesh, if this is not the case it proceeds to
+   * mesh it automatically. We remark that the automatic mesh generation in
+   * OpenCASCADE takes care only of the geometrical resemblance between the
+   * shape and the mesh, to control the shape and regularity of the triangles
+   * you should use other meshing softwares. The two double deflection and
+   * angular deflection select the accuracy of the created triangulation with
+   * respect to the orginal topological shape. The bool sew_different_faces
```
Comment (bangerth) [inline include/deal.II/opencascade/utilities.h]:
```suggestion
   * respect to the orginal topological shape. The argument `sew_different_faces`
```

File: source/opencascade/utilities.cc [hunk 2]
```diff
@@ -267,6 +276,67 @@ namespace OpenCASCADE
...
 
+
+  TopoDS_Shape
+  read_STL(const std::string &filename)
+  {
+    StlAPI_Reader reader;
+    TopoDS_Shape  shape;
+    reader.Read(shape, filename.c_str());
+    return shape;
+  };
```
Comment (bangerth) [inline source/opencascade/utilities.cc]:
```suggestion
  }
```

File: source/opencascade/utilities.cc [hunk 3]
```diff
@@ -267,6 +276,67 @@ namespace OpenCASCADE
...
+    std::vector<TopoDS_Face>   faces;
+    OpenCASCADE::extract_geometrical_shapes(shape, faces, edges, vertices);
+    for (unsigned int i = 0; i < faces.size(); ++i)
+      {
+        Handle(Poly_Triangulation) theTriangulation =
+          BRep_Tool::Triangulation(faces[i], Loc);
+        if (theTriangulation.IsNull())
+          mesh_is_present = false;
+      }
+    // std::cout<<mesh_is_present<<std::endl;
```
Comment (bangerth) [inline source/opencascade/utilities.cc]:
```suggestion

```

File: source/opencascade/utilities.cc [hunk 4]
```diff
@@ -267,6 +276,67 @@ namespace OpenCASCADE
...
+  write_STL(const TopoDS_Shape &shape,
+            const std::string & filename,
+            const double        deflection,
+            const double        angular_deflection,
+            const bool          sew_different_faces)
+  {
+    // Check if mesh is present
+    bool mesh_is_present = true;
+    Handle(StlMesh_Mesh) stl_mesh;
+    TopoDS_Shape               shape_to_be_written = shape;
```
Comment (bangerth) [inline source/opencascade/utilities.cc]: Move this variable declaration to just above where you need it first.

File: source/opencascade/utilities.cc [hunk 5]
```diff
@@ -267,6 +276,67 @@ namespace OpenCASCADE
...
+    TopLoc_Location            Loc;
+    std::vector<TopoDS_Vertex> vertices;
+    std::vector<TopoDS_Edge>   edges;
+    std::vector<TopoDS_Face>   faces;
+    OpenCASCADE::extract_geometrical_shapes(shape, faces, edges, vertices);
+    for (unsigned int i = 0; i < faces.size(); ++i)
```
Comment (bangerth) [inline source/opencascade/utilities.cc]: Move the declaration of `mesh_is_present` to just above the loop here.

File: source/opencascade/utilities.cc [hunk 6]
```diff
@@ -267,6 +276,67 @@ namespace OpenCASCADE
...
+        double  x_min, y_min, z_min, x_max, y_max, z_max;
+        Bnd_Box bnd_box;
+        BRepBndLib::Add(shape_to_be_written, bnd_box);
+        bnd_box.Get(x_min, y_min, z_min, x_max, y_max, z_max);
+        BRepMesh_FastDiscret mesh_fd(deflection, angular_deflection, bnd_box);
+        mesh_fd.Perform(shape_to_be_written);
+      }
+    // Write out stl mesh to file
+    StlAPI_Writer writer;
+    auto          error = writer.Write(shape_to_be_written, filename.c_str());
```
Comment (bangerth) [inline source/opencascade/utilities.cc]:
```suggestion
    const auto          error = writer.Write(shape_to_be_written, filename.c_str());
```

File: tests/opencascade/stl_creation_single_patch.cc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Read a file in iges format, and write it out again in the same
```
Comment (bangerth) [inline tests/opencascade/stl_creation_single_patch.cc]:
```suggestion
// Read a file in iges format, and write it out again in STL
```

File: tests/opencascade/stl_creation_watertight.cc [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Read a file in iges format, and write it out again in the same
+// format.
```
Comment (bangerth) [inline tests/opencascade/stl_creation_watertight.cc]: Update the comment to say what's specific about this test. Also mention that you're outputting in STL.

File: tests/opencascade/stl_refinement.cc [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Read a file in iges format, and write it out again in the same
+// format.
```
Comment (bangerth) [inline tests/opencascade/stl_refinement.cc]: same here

File: tests/opencascade/stl_write.cc [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Read a file in iges format, and write it out again in the same
+// format.
```
Comment (bangerth) [inline tests/opencascade/stl_write.cc]: and here

**Comment by nicola-giuliani** [comment]
Thank you @bangerth  for the precious comments, I've updated the rebased branch and squashed the commits

File: include/deal.II/opencascade/utilities.h [hunk 11]
```diff
@@ -128,6 +128,35 @@ namespace OpenCASCADE
...
+
+  /**
+   * Write the given topological shape into an STL file. In order to do so the
+   * shape must contain a mesh structure, the function checks if all the faces
+   * of the shape have an attached mesh, if this is not the case it proceeds to
+   * mesh it automatically. We remark that the automatic mesh generation in
+   * OpenCASCADE takes care only of the geometrical resemblance between the
+   * shape and the mesh, to control the shape and regularity of the triangles
+   * you should use other meshing softwares. The two double `deflection` and
+   * `angular deflection` select the accuracy of the created triangulation with
```
Comment (bangerth) [inline include/deal.II/opencascade/utilities.h]:
```suggestion
   * `angular_deflection` select the accuracy of the created triangulation with
```

File: include/deal.II/opencascade/utilities.h [hunk 12]
```diff
@@ -128,6 +128,35 @@ namespace OpenCASCADE
...
+  read_STL(const std::string &filename);
+
+  /**
+   * Write the given topological shape into an STL file. In order to do so the
+   * shape must contain a mesh structure, the function checks if all the faces
+   * of the shape have an attached mesh, if this is not the case it proceeds to
+   * mesh it automatically. We remark that the automatic mesh generation in
+   * OpenCASCADE takes care only of the geometrical resemblance between the
+   * shape and the mesh, to control the shape and regularity of the triangles
+   * you should use other meshing softwares. The two double `deflection` and
```
Comment (bangerth) [inline include/deal.II/opencascade/utilities.h]:
```suggestion
   * you should use other meshing softwares. The two arguments `deflection` and
```

File: source/opencascade/utilities.cc (hunk 4 above)
Comment (bangerth) [reply source/opencascade/utilities.cc]: How about this comment and the one below?

File: include/deal.II/opencascade/utilities.h (hunk 11 above)
Comment (nicola-giuliani) [reply include/deal.II/opencascade/utilities.h]: I missed them the first time.

**Comment by nicola-giuliani** [comment]
The comments were some leftovers I used to code the different parts, re-reading the code it appears clear what each part is doing so I erased them

File: source/opencascade/utilities.cc [hunk 13]
```diff
@@ -267,6 +276,64 @@ namespace OpenCASCADE
...
+        Handle(Poly_Triangulation) theTriangulation =
+          BRep_Tool::Triangulation(faces[i], Loc);
+        if (theTriangulation.IsNull())
+          mesh_is_present = false;
+      }
+    if (!mesh_is_present)
+      {
+        if (sew_different_faces)
+          {
+            const double          tolerance = 1e-06;
```
Comment (luca-heltai) [inline source/opencascade/utilities.cc]: Can you make this tolerance also an argument of the function (say `sewer_tolerance`)

Comment (nicola-giuliani) [reply source/opencascade/utilities.cc]: yep, I will recompile to check that everything works fine. Then I will rebase and push

Comment (nicola-giuliani) [reply source/opencascade/utilities.cc]: done

**Comment by nicola-giuliani** [comment]
should have done that ;)

**Comment by masterleinad** [comment]
There are some more issues with this pull request, see https://cdash.43-1.org/viewBuildError.php?onlydeltap&buildid=2671.

**Comment by luca-heltai** [comment]
I suspect that this is due to an old version of opencascade installed on  the tester?
What version is that? In this case, we should wrap these two functions into #ifdefs...

**Comment by masterleinad** [comment]
In https://cdash.43-1.org/viewBuildError.php?type=1&buildid=2702, it says version 7.3.0.

**Comment by tamiko** [comment]
The following version is installed on the tester:
```
[I] sci-libs/opencascade
     Available versions:  (7.3.0) (~)7.3.0{tbz2}
       {debug doc examples ffmpeg freeimage gl2ps gles2 java +tbb +vtk ELIBC="FreeBSD"}
     Installed versions:  7.3.0(7.3.0){tbz2}(11:38:18 PM 08/12/2019)(doc ffmpeg tbb vtk -debug -examples -freeimage -gl2ps -gles2 -java ELIBC="-FreeBSD")
     Homepage:            https://www.opencascade.com/
     Description:         Development platform for CAD/CAE, 3D surface/solid modeling and data exchange

```

**Comment by tamiko** [comment]
it looks like that #9019 might not fix the compile errors on the tester.

@masterleinad @luca-heltai How shall we proceed?

**Comment by masterleinad** [comment]
Revert until we find a solution?

**Comment by tamiko** [comment]
@luca-heltai @masterleinad It looks like that the error codes in question are present in the opencascade code base until version 7.1.0 [1] and have been removed in version 7.2.0 [2] onwards. I currently cannot clone the opencascade git repository to confirm my suspicion.

[1] https://www.opencascade.com/doc/occt-7.1.0/refman/html/_stl_a_p_i___error_status_8hxx.html
[2] https://www.opencascade.com/doc/occt-7.2.0/refman/html/

**Comment by masterleinad** [comment]
There is also an error saying
```
error: no matching constructor for initialization of 'BRepMesh_FastDiscret'
/usr/lib64/opencascade-7.3.0/ros/include/opencascade/BRepMesh_FastDiscret.hxx:116:19: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
[...]
```

**Comment by dangars** [comment]
> @luca-heltai @masterleinad It looks like that the error codes in question are present in the opencascade code base until version 7.1.0 [1] and have been removed in version 7.2.0 [2] onwards. I currently cannot clone the opencascade git repository to confirm my suspicion.

@tamiko  I don't know where git clone the official OpenCascade version. I see in the [website](https://dev.opencascade.org/index.php?q=home/resources) that you need a ssh key, but OpenCascade 7.3 is available in Debian. Tomorrow I'll try to get the sources from Debian.

**Comment by tamiko** [comment]
We have temporarily reverted the merge in order to have a functional testsuite again. Let's work on getting this merged again soon!

**Comment by tamiko** [comment]
@nicola-giuliani @luca-heltai *ping* May I ask what version of opencascade this was based on?

@dangars I see that 7.4 has been released as well. We should ideally cover a wide range of releases - I'd say 6.9/7.0 (or so) onwards.

**Comment by nicola-giuliani** [comment]
@tamiko it was based on the Darwin-Catalina packages which is based on oce-0.18.3 (and I think on the spack installation procedure), which is based on OpenCASCADE 6.9. From the OCE github page the last released is 0.18.3, 11 days ago two new releases of OpenCASCADE were added on the OCE Release page but I don't think they are in OCE yet.  @luca-heltai @tamiko do you agree?

**Comment by nicola-giuliani** [comment]
I have changed the used function to BRepMesh_IncrementalMesh that should be used from 6.9.0 to 7.4.0. I still don't get how to check if the version is > than 6.8.0.

**Comment by masterleinad** [comment]
> I have changed the used function to BRepMesh_IncrementalMesh that should be used from 6.9.0 to 7.4.0.

You should create a new pull request for adding this functionality back in.

>  I still don't get how to check if the version is > than 6.8.0.

Can't you just use  
```
#  if (OCC_VERSION_MAJOR >= 6 && OCC_VERSION_MINOR >=9) 
```
(or similar)?

**Comment by nicola-giuliani** [comment]
Thanks @masterleinad , PR coming very soon

**Comment by bangerth** [comment]
On 11/11/19 9:00 AM, Daniel Arndt wrote:
> |# if (OCC_VERSION_MAJOR >= 6 && OCC_VERSION_MINOR >=9) |

Well, that doesn't match 7.0, 7.1, ..., 7.8 :-) But there are obvious 
generalizations :-)

**Pull request merged**